File: obiluascripts/cleanacgt.lua
-- ============================================================
-- cleanacgt.lua
-- OBITools4 obiscript — split sequences into [ACGTacgt]+ segments.
--
-- Non-ACGT characters (N, IUPAC ambiguity codes, gaps, …) act as
-- split points.  Each contiguous run of pure ACGT bases is emitted
-- as a separate output sequence.  The position within the parent
-- sequence is preserved via subsequence(), so quality scores and
-- annotations are carried over when available.
--
-- Usage:
--   obiscript -S /app/obiluascripts/cleanacgt.lua <input.fasta.gz>
-- ============================================================

function worker(sequence)
    local seq_str = sequence:sequence()
    local slice   = BioSequenceSlice.new()
    local pos     = 1

    while pos <= #seq_str do
        local s, e = seq_str:find("[ACGTacgt]+", pos)
        if not s then break end
        -- subsequence() is 0-based, end exclusive
        local sub = sequence:subsequence(s - 1, e)
        slice:push(sub)
        pos = e + 1
    end

    return slice
end

File: src/skimindex_py/skimindex/processing/cleanacgt.py
"""
skimindex.processing.cleanacgt — atomic 'cleanacgt' processing type.

Splits sequences into pure-ACGT segments by cutting on any non-ACGT
character (N, IUPAC ambiguity codes, gaps, …) using cleanacgt.lua.
Output kind: STREAM (chainable via Data → Data interface).
"""

from collections.abc import Callable

from skimindex.processing import OutputKind, processing_type
from skimindex.processing.data import Data, stream_data, to_stream_command
from skimindex.unix.obitools import obiscript

CLEANACGT_LUA = "/app/obiluascripts/cleanacgt.lua"


@processing_type(output_kind=OutputKind.STREAM, output_filename="cleanacgt.fasta")
def cleanacgt(params: dict) -> Callable[[Data], Data]:
    """Split sequences into contiguous [ACGTacgt]+ segments.

    Non-ACGT characters act as split points; each resulting pure-ACGT
    run is emitted as a separate sequence.  No parameters required.
    """
    def run(input_data: Data) -> Data:
        cmd = to_stream_command(input_data)
        return stream_data(
            cmd | obiscript(CLEANACGT_LUA),
            format="fasta",
            subdir=input_data.subdir,
            per_species=input_data.per_species,
        )

    return run
